Simplify swappiness check

hqhq opened the pull request:
As suggested in https://github.com/docker/docker/pull/14004/files#r34022527

The concern there is we can't differentiate whether user explicitly
asked for an invalid value of -1 or he did not specify anything.

I don't think this would be a problem, because:
- like all other default values like zero, we can't differentiate
  user specify it or not, most of which, zeros are also invalid, so
  default is default, we show these default values in help info,
  so users would know if they set value as default, it'll be like
  they set nothing.
- we can't do this kind of string check in REST api request, so
  it'll make the behave different from docker command and RESTapi.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

estesp commented:
The `TestRunWithSwappinessInvalid` test is broken as it expected passing "-1" to fail.  That test can be changed, but we do need to verify that users are told via documentation, and will expect "-1" to be regarded as "don't modify `memory.swappiness` from default/parent value"

hqhq commented:
@estesp Thanks, updated. I added some docs about "-1" value, I didn't change all the places where claim the invalid value is 0-100, because I think it's not worth mentioning, and it's not recommended to use this "-1" value, just for users who happen to use this value and found no error, they can seek for the answer from somewhere.

cpuguy83 commented:
You have a failing swappiness test.

cpuguy83 commented:
Also for other maintainers:

Libcontainer only sets a swappiness value when the value is between `0-100`, otherwise it does nothing, which means it would just inherit the parent cgroup value.
https://github.com/docker/libcontainer/blob/master/cgroups/fs/memory.go#L74-L77

hqhq commented:
@cpuguy83 Looks like it's Jenkins' problem.

cpuguy83 commented on docs/reference/run.md:
```diff
@@ -617,7 +617,13 @@ and require killing system processes to free memory.
 To set this percentage for a container, specify a `--memory-swappiness` value
 between 0 and 100. A value of 0 turns off anonymous page swapping. A value of
-100 sets all anonymous pages as swappable.
+100 sets all anonymous pages as swappable. By default, if you are not using
```
I would change all the `we`'s here to `Docker`, maybe `Docker Engine`... I'm sure the docs team would have an opinion there.

cpuguy83 commented:
LGTM

estesp commented:
LGTM, @moxiegirl or @thaJeztah, this needs docs review.  Thanks!

moxiegirl commented on docs/reference/run.md:
```diff
@@ -617,7 +617,13 @@ and require killing system processes to free memory.
+100 sets all anonymous pages as swappable. By default, if you are not using
+`--memory-swapiness`, memory swappiness value will be inherited from parent.
+
+Note that since 0 is an effective value, so we take -1 as the default value
```
I would leave this paragraph out.  We say valid values are between 0-100 why do we need to mention -1 case?

hqhq replied:
@moxiegirl Updated.

moxiegirl commented:
Thanks @hqhq LGTM @thaJeztah for the merge

estesp commented on docs/reference/run.md:
```diff
@@ -617,7 +617,8 @@ and require killing system processes to free memory.
 between 0 and 100. A value of 0 turns off anonymous page swapping. A value of
-100 sets all anonymous pages as swappable.
+100 sets all anonymous pages as swappable. By default, if you are not using
+`--memory-swapiness`, memory swappiness value will be inherited from parent.
```
sorry but since we haven't merged yet let's fix the typo in the flag-name: `--memory-swappiness`.  And let's add the word "from **the** parent." at the end of the sentence to read a little better.

thaJeztah replied:
Good catch, sgtm!

@hqhq could you address this, then we can merge

hqhq replied:
Done, thanks.

hqhq committed:
Simplify swappiness check

As suggested in https://github.com/docker/docker/pull/14004/files#r34022527

The concern there is we can't differentiate whether user explicitly
asked for an invalid value of -1 or he did not specify anything.

I don't think this would be a problem, because:
 - like all other default values like zero, we can't differentiate
   user specify it or not, most of which, zeros are also invalid, so
   default is default, we show these default values in help info,
   so users would know if they set value as default, it'll be like
   they set nothing.
 - we can't do this kind of string check in REST api request, so
   it'll make the behave different from docker command and RESTapi.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

thaJeztah commented:
LGTM, thanks @hqhq!
